[CBRD-26058] Fix error in CASE expression with logical type

HyunukLee opened the pull request:
<http://jira.cubrid.org/browse/CBRD-26058>

### Purpose

특정 상황에서 SQL CASE문이 logical 타입일 경우 발생하는 에러 해결

### Implementation

expression이 logical 타입인 경우에 대해 처리하는 코드에서 CASE문을 처리하지 않으므로, CASE문이 logical 타입이더라도 기존에 처리하던대로 수행하도록 수정

### Remarks

CASE문이 logical 타입이 되는 상황은 다음과 같습니다.
THEN절 또는 ELSE절 중 하나는 비교식의 형태, 다른 하나는 '= NULL' 등의 조건으로 인해 결과가 NULL인 형태일 때 입니다.

HyunukLee committed:
[fix] PT_TYPE_LOGICAL 타입의 PT_CASE 노드 처리 추가

- 문제: CASE 문의 타입이 PT_TYPE_LOGICAL인 경우에 대한 처리가 누락되어 있었음
- 해결: PT_TYPE_LOGICAL 타입일 경우 예외적으로 기존 CASE 처리 로직을 타도록 수정함

HyunukLee committed:
Merge branch 'CUBRID:develop' into RND-2414

youngjinj commented on src/parser/xasl_generation.c:
```diff
@@ -7545,7 +7545,7 @@ pt_to_regu_variable (PARSER_CONTEXT * parser, PT_NODE * node, UNBOX unbox)
 
       if (node != NULL && node->type_enum == PT_TYPE_LOGICAL
-	  && (node->node_type == PT_EXPR || node->node_type == PT_VALUE))
+	  && (node->node_type == PT_EXPR || node->node_type == PT_VALUE) && node->info.expr.op != PT_CASE)
```
```suggestion
	  && ((node->node_type == PT_EXPR && node->info.expr.op != PT_CASE) || node->node_type == PT_VALUE))
```
비슷한 비교를 같이 하면 좋을 것 같습니다.

HyunukLee replied:
6fcdd7c에서 수정했습니다.

HyunukLee committed:
[style] EXPR 노드에 대한 비교 묶기

노드가 EXPR인지, PT_CASE인지에 대한 비교를 묶음으로써 가독성 향상

Co-authored-by: Joo Youngjin <68979537+youngjinj@users.noreply.github.com>

youngjinj commented on src/parser/xasl_generation.c:
```diff
@@ -7545,7 +7545,7 @@ pt_to_regu_variable (PARSER_CONTEXT * parser, PT_NODE * node, UNBOX unbox)
 
       if (node != NULL && node->type_enum == PT_TYPE_LOGICAL
-	  && (node->node_type == PT_EXPR || node->node_type == PT_VALUE))
+	  && ((node->node_type == PT_EXPR && node->info.expr.op != PT_CASE) || node->node_type == PT_VALUE))
```
이 코드 대신에
`pt_common_type_op`를 아래처럼 변경하는 것은 어떤가요?

```sql
    /*
     * true + true must not be logical.
     * Same goes for true*true, (i or j)+(i and j) etc.
     * Basic rule: if both operands are logical but the operation is not logical,
     * then the resulting type MUST be integer. Otherwise strange things happen
     * while generating xasl: predicate expressions with wrong operators.
     */
(-) if (t1 == PT_TYPE_LOGICAL && t2 == PT_TYPE_LOGICAL && !pt_is_operator_logical (op))
(+) if (result_type == PT_TYPE_LOGICAL && !pt_is_operator_logical (op))
      {
        result_type = PT_TYPE_INTEGER;
      }
```

주석을 보면 XASL을 생성할 때 문제가 발생하지 않게 하기 위해서 이 코드가 추가되었습니다.
`t1`과 `t2` 중 하나가 `PT_TYPE_LOGICAL`이 아닐 때 이 코드를 우회하고 있는 것으로 보입니다.

두 타입이 모두 `PT_TYPE_LOGICAL`이거나 하나라도 `PT_TYPE_NULL`인 경우의 `result_type`은 `PT_TYPE_LOGICAL`이 됩니다.
하나라도 `PT_TYPE_MAYBE`인 경우에는 `result_type`도 `PT_TYPE_MAYBE`가 되기 때문에 문제가 되지 않습니다.

Syntax를 체크할 때 `parser_make_expression`를 보면 `pt_is_operator_logical`를 만족하는 경우에만 `type_enum`을 `PT_TYPE_LOGICAL`로 설정하고 있습니다.
그래서 `result_type`이 `PT_TYPE_LOGICAL`일 때 `pt_is_operator_logical`를 만족하는지 검사하는 것이 타당하다고 생각합니다.

shparkcubrid replied:
'node->info.expr.op != PT_CASE' 대신에 pt_is_operator_logical()으로 체크하는 건 어떨까요? PT_CASE만 제외하는 것보다는 의미가 있어보입니다.

HyunukLee replied:
영진님이 제안하신 방식대로 수정하면,
PT_CASE에만 국한되지 않고, node_type이 PT_TYPE_LOGICAL이지만 pt_is_operator_logical()을 만족하지 않는 모든 경우를 포괄할 수 있어 합리적이라고 생각합니다.

또한 체크 시점의 차이만 있을 뿐 세훈님이 말씀하신 내용도 포함되는 셈입니다.

해당 내용 반영하여 수정하겠습니다.

HyunukLee replied:
7896686에서 수정하였습니다.

youngjinj replied:
저는 7896686 의 변경사항을 제안드린 것이라서 Resolve 해도 괜찮습니다.
@shparkcubrid 님의 동의를 통해서 Resolve 여부를 판단해주시면 될 것 같습니다.

HyunukLee committed:
[refactor] pt_common_type_op에서 체크하도록 변경

- 변경 : 피연산자가 모두 PT_TYPE_LOGICAL일 때만 op가 logcial인지 확인했었는데, 결과 타입이 PT_TYPE_LOGICAL인 경우로 확장
- 목적
1. 기존 반영 사항이었던 pt_to_regu_variable에서 노드타입이 PT_TYPE_LOGICAL인 경우에 대한 조건문이 노드타입에 대해서만 체크하도록 일관성 유지
2. CASE문뿐 아니라 다른 경우에 대한 처리가 가능하도록 확장

HyunukLee commented:
/run shell

HyunukLee commented:
/run all

HyunukLee commented:
/run all

HyunukLee commented:
/run shell

HyunukLee committed:
Merge branch 'CUBRID:develop' into RND-2414

HyunukLee commented:
/run shell

HyunukLee commented:
/run all

HyunukLee committed:
Merge branch 'CUBRID:develop' into RND-2414

HyunukLee commented:
/run shell

HyunukLee commented:
/run shell

HyunukLee commented:
/run medium sql
